gnomad/utils/constraint.py

```diff
@@ -700,3 +701,377 @@ def get_all_pop_lengths(
+        variants.
+    :param annotation_name: Annotation name used for constraint metrics to distinguish
+        mutation types.
+    :param keys: The keys of the output Table. Default is ('gene', 'transcript',
+        'canonical').
+    :param pops: List of populations used to compute constraint metrics. Default is ().
+    :return: Table with observed:expected ratio.
+    """
+    # Create aggregators that sum the number of observed variants, possible variants,
+    # and expected variants and compute observed:expected ratio.
+    agg_expr = {
+        f"obs_{annotation_name}": hl.agg.sum(ht.observed_variants),
+        f"exp_{annotation_name}": hl.agg.sum(ht.expected_variants),
+        f"oe_{annotation_name}": hl.agg.sum(ht.observed_variants)
+        / hl.agg.sum(ht.expected_variants),
+        f"possible_{annotation_name}": hl.agg.sum(ht.possible_variants),
+    }
+
+    # Create an aggregator that sums the mutation rate.
+    if annotation_name != "mis_pphen":
+        agg_expr[f"mu_{annotation_name}"] = hl.agg.sum(ht.mu)
+
+    # Create aggregators that sum the number of observed variants
+    # and expected variants for each population if pops` is specified.
+    for pop in pops:
+        agg_expr[f"exp_{annotation_name}_{pop}"] = hl.agg.array_sum(
+            ht[f"expected_variants_{pop}"]
+        )
+        agg_expr[f"obs_{annotation_name}_{pop}"] = hl.agg.array_sum(
+            ht[f"downsampling_counts_{pop}"]
+        )
+    return ht.group_by(*keys).aggregate(**agg_expr)
+
+
+def compute_all_pLI_scores(
+    ht: hl.Table,
+    annotation_name: str,
+    pops: Tuple[str] = (),
+    keys: Tuple[str] = ("gene", "transcript", "canonical"),
+    calculate_pop_pLI=False,
+    n_ds_to_skip: int = 8,
+) -> hl.Table:
+    """
+    Compute the pLI scores for pLoF variants.
+
+    :param ht: Input Table with pLoF variants.
+    :param annotation_name: Annotation name used for constraint metrics to distinguish
+        mutation types.
+    :param pops: Populations that need to compute pLI scores.
+    :param keys: The keys of the output Table. Default is ('gene', 'transcript',
+        'canonical').
+    :param calculate_pop_pLI: Whether to compute the pLI scores for each population.
+        Default is False.
+    :param n_ds_to_skip: The number of small downsamplings to skip when calculating pLI scores for populations. Default is 8.
+    :return: Table with pLI, pNull, and pRec scores.
+    """
+    # Filter to only variants with expected variant counts larger than 0.
+    ht = ht.filter(ht[f"exp_{annotation_name}"] > 0)
+
+    # Calculate pLI scores for each population if specified.
+    if calculate_pop_pLI:
+        pop_lengths = get_all_pop_lengths(ht, pops, f"obs_{annotation_name}_{pop}")
+        for pop_length, pop in pop_lengths:
+            logger.info(f"Calculating pLI for {pop}...")
+            plis = []
+            for i in range(n_ds_to_skip, pop_length):
```

Comment:
As Konrad said, it may be changed in v4, so I made it as a parameter. We may need to think a better way to define the number.

Reply:
For now we are not implementing the pop downsampling pLI, and a TODO has been added to the constraint code to add this if necessary.